libp2p/autotls/autotls.nim

```diff
@@ -0,0 +1,249 @@
+    if txt.len > 0 and txt[0] != "not set yet" and ip4.len > 0:
+      return true
+    await sleepAsync(1.seconds)
+
+  return false
+
+proc new*(
+    T: typedesc[AutoTLSManager],
+    rng: ref HmacDrbgContext = newRng(),
+    acmeAccount: ref ACMEAccount = nil,
+    dnsResolver: DnsResolver = DnsResolver.new(DefaultDnsServers),
+): AutoTLSManager =
+  T(
+    rng: rng,
+    managerLoop: nil,
+    cert: Opt.none(TLSCertificate),
+    certExpiry: Opt.none(Moment),
+    certReady: newAsyncEvent(),
+    acmeAccount: acmeAccount,
+    httpSession: HttpSessionRef.new(),
+    dnsResolver: dnsResolver,
+    peerInfo: Opt.none(PeerInfo),
+    bearerToken: Opt.none(string),
+    renewCheckTime: DefaultRenewCheckTime,
+    renewBufferTime: DefaultRenewBufferTime,
+  )
+
+method issueCertificate(
+    self: AutoTLSManager
+): Future[void] {.base, async: (raises: [AutoTLSError, CancelledError]).} =
+  trace "Issuing new certificate"
+  var peerInfo: PeerInfo
+  try:
+    peerInfo = self.peerInfo.get
+  except:
+    raise newException(AutoTLSError, "Cannot issue new certificate: peerInfo not set")
+
+  # generate autotls domain string: "*.{peerID}.libp2p.direct"
+  let base36PeerId = encodePeerId(peerInfo.peerId)
+  let baseDomain = base36PeerId & "." & AutoTLSDNSServer
+  let domain = "*." & baseDomain
+
+  debug "Requesting ACME challenge"
+  let (dns01Challenge, finalizeURL, orderURL) =
+    await self.acmeAccount.requestChallenge(@[domain])
+
+  let keyAuthorization = base64UrlEncode(
+    @(
+      sha256.digest(
+        (
+          dns01Challenge.getJSONField("token").getStr & "." &
+          thumbprint(self.acmeAccount.key)
+        ).toByteSeq
+      ).data
+    )
+  )
+
+  var strMultiaddresses: seq[string] = @[]
+  for ma in peerInfo.addrs:
+    strMultiaddresses.add($ma)
+
+  let payload = %*{"value": keyAuthorization, "addresses": strMultiaddresses}
+  let registrationURL = AutoTLSBroker & "/v1/_acme-challenge"
+
+  debug "Sending challenge to AutoTLS broker"
+  var response: HttpClientResponseRef
+  var bearerToken: string
+  if self.bearerToken.isSome:
+    (bearerToken, response) = await peerIdAuthSend(
+      registrationURL,
+      self.httpSession,
+      peerInfo,
+      payload,
+      bearerToken = self.bearerToken,
+    )
+    discard bearerToken
+  else:
+    # authenticate, send challenge and save bearerToken for future requests
+    (bearerToken, response) =
+      await peerIdAuthSend(registrationURL, self.httpSession, peerInfo, payload)
+    self.bearerToken = Opt.some(bearerToken)
+
+  # no need to do anything from this point forward if there are not public ip addresses on host
+  var hostPrimaryIP: IpAddress
+  try:
+    hostPrimaryIP = getPrimaryIPAddr()
+    if not isPublicIPv4(hostPrimaryIP):
+      raise newException(AutoTLSError, "Host does not have a public IPv4 address")
+  except Exception:
+    raise newException(AutoTLSError, "Failed to get primary IP address for host")
```

Comment:
code should not catch `Exception` because it is root of `CatchableError` and `Defect`. 
`Defect` is catchable, but code shouldn't catch it unless it is deliberately done for some reason.
therefore in all our code we catch only `CatchableError`.

```
Exception (root of all exceptions)
├── CatchableError (recoverable, can be caught with `except`)
│  
└── Defect (programming bugs, NOT catchable with `except`)
    ├── AssertionDefect
    ├── NilAccessDefect
    └── IndexDefect, etc.
```

Reply:
Thx for the tip!

Reply:
Changing this makes nim 1.6 fail